Dockerfile

```diff
@@ -1,15 +1,32 @@
-FROM rust:1.74 as build-env
+FROM rust:1.75 as build-env
 RUN rustup target add x86_64-unknown-linux-musl
 WORKDIR /app
 COPY Cargo.toml /app
 COPY Cargo.lock /app
-RUN mkdir /app/src
-RUN echo "fn main() {}" > /app/src/main.rs
+RUN mkdir -p /app/crates/trippy/src
+RUN mkdir -p /app/crates/trippy-core/src
+RUN mkdir -p /app/crates/trippy-dns/src
+RUN mkdir -p /app/crates/trippy-privilege/src
+COPY crates/trippy/Cargo.toml /app/crates/trippy/Cargo.toml
+COPY crates/trippy-core/Cargo.toml /app/crates/trippy-core/Cargo.toml
+COPY crates/trippy-dns/Cargo.toml /app/crates/trippy-dns/Cargo.toml
+COPY crates/trippy-privilege/Cargo.toml /app/crates/trippy-privilege/Cargo.toml
+
+# dummy build to cache dependencies
+RUN echo "fn main() {}" > /app/crates/trippy/src/main.rs
+RUN touch /app/crates/trippy-core/src/lib.rs
+RUN touch /app/crates/trippy-dns/src/lib.rs
+RUN touch /app/crates/trippy-privilege/src/lib.rs
 RUN cargo build --release --target=x86_64-unknown-linux-musl
-COPY src /app/src
+
+# copy the actual application code and build
+COPY crates/trippy/src /app/crates/trippy/src
+COPY crates/trippy-core/src /app/crates/trippy-core/src
+COPY crates/trippy-dns/src /app/crates/trippy-dns/src
+COPY crates/trippy-privilege/src /app/crates/trippy-privilege/src
 COPY trippy-config-sample.toml /app
 COPY README.md /app
-COPY LICENSE /app
+RUN cargo clean --release --target=x86_64-unknown-linux-musl -p trippy-core -p trippy-dns -p trippy-privilege
```

Comment:
I don't know why this is needed, but it is.

Reply:
the dependency caching step creates empty dummy libraries with the same names as the trippy crates. then, when you copy the production code in, they try (and fail) to import modules from those empty dummy libraries -- hence why you need to delete the build artifacts.

Reply:
@wgreenberg I'd have expected that to work though?  The final build step (after copying in the real code) should rebuild the library crates as their src code has changed but it seems it doesn't "notice" that.

This approach certainly works for a single crate (without a virtual workspace) as that is what it does today: https://github.com/fujiapple852/trippy/blob/74c2c6a5a1d5f019858de78c83a7f77c8dbc0b4d/Dockerfile#L7-L13

So I can only assume a bug?

I did try disabling incremental compilation (`CARGO_INCREMENTAL=0`) in case it was the cause but that didn't seem to help.

Reply:
Resolving this for now as it does not block merging, but it would be nice to understand what the issue is here and either fix it our report a bug.